### DISCUSSION TITLE
데이터베이스의 여러 조인

### DISCUSSION THREAD

**Discussion by sujin-park0607** [24tngus/CS_STUDY#34] [Database]
# 조인

> 목차
> 1. 조인 종류
> 2. 조인 수행 방식

🔗 조인 학습 참고tool 
https://sql-joins.leopard.in.ua/

## 1. 조인이란?
하나의 테이블이 아닌 두 개 이상의 테이블을 묶어서 하나의 결과물을 만드는 것 MySQL에서는 Join이라는 쿼리로, MongoDB에서는 lookup이라는 쿼리로 처리할 수 있음

> MongoDB를 사용할 때 lookup은 되도록 사용하지 말아야한다. → MongoDB는 조인연산(lookup)에 대해 관계형 데이터베이스보다 성능이 떨어진다.

![image](https://github.com/24tngus/CS_STUDY/assets/75667075/5efeb61a-3334-4cb1-8df1-4f15e3202aee)

**내부조인(inner join):** 왼쪽 테이블과 오른쪽 테이블의 두 행이 모두 일치하는 행이 있는 부분만 표기된다.

```
SELECT * FROM TableA A
INNER JOIN TableB B ON
A.key = B.key
```
![image](https://github.com/24tngus/CS_STUDY/assets/75667075/15163ba6-5a8d-41fa-be00-1ea908ededf8)

**자연 조인(Natural Join):** 내부 조인에 속하며 두 테이블에서 동일한 컬럼명을 갖는 컬럼은 모두 조인이 된다.

**왼쪽조인(lefter outer join):** 왼쪽 테이블의 모든 행이 결과 테이블에 표기된다.
```
SELECT * FROM TableA A
LEFT JOIN TableB B ON
A.key = B.key
```
![image](https://github.com/24tngus/CS_STUDY/assets/75667075/3511b9fb-2bb7-4240-af0e-56009c5f2368)

**오른쪽 조인(right outer join):** 오른쪽 테이블의 모든 행이 결과 테이블에 표기된다.
```
SELECT * FROM TableA A
RIGHT JOIN TableB B ON
A.key = B.key
```
![image](https://github.com/24tngus/CS_STUDY/assets/75667075/7eaa8296-8a36-4d3d-a0de-1a0878206986)

**합집합 조인(full outer join):** 두 개의 테이블을 기반으로 조인 조건에 만족하지 않는 행까지 모두 표기된다.
```
SELECT * FROM TableA A
FULL OUTER JOIN TableB B ON
A.key = B.key
```
![image](https://github.com/24tngus/CS_STUDY/assets/75667075/32930aaa-156a-400d-95a4-dca3aef7204c)

**크로스 조인(Cross Join):** 두 테이블에서 곱집합을 반환한다.
```
SELECT *
FROM EMPLOYEE 
CROSS JOIN DEPARTMENT;
```
![image](https://github.com/24tngus/CS_STUDY/assets/75667075/3b2c2f98-89d3-4726-8f9f-47a5a72fd5b6)

**셀프 조인(Self Join):** 자기 자신과 조인한다.
```
SELECT
A.NAME, B.AGE
FROM TABLE_A A, TABLE_A B
```
![image](https://github.com/24tngus/CS_STUDY/assets/75667075/1a7027b9-570d-4077-8e65-a48ad5931508)

## 2. 조인 수행 방식
테이블 조인 방식은 테이블의 각 로우를 매핑 시 어떤 메커니즘을 사용하는냐에 따라 조인을 구분한다.

**중첩루프조인(Nested Loop Join)**

- (Nested Loop Join)은 중첩 for문과 같은 원리, 2개의 리스트를 완전탐색한다고 생각하면 된다. 
- 랜덤 접근에 대한 비용이 많이 증가하므로 대용량의 테이블에서는 사용하지 않는다.
![image](https://github.com/24tngus/CS_STUDY/assets/75667075/637ca93e-a1b2-4d63-8c1d-d049da48860b)

**정렬병합조인(Sort Merge Join)**

- 각각의 테이블을 조인할 필드 기준으로 정렬하고 정렬이 끝난 이후에 조인 작업을 수행하는 조인이다.
- 적절한 인덱스가 없고 대용량의 테이블들을 조인하고, 조인조건으로 >,<등 범위 비교 연산자가 있을 때 사용한다.

**해시 조인(Hash Join)**

- 해시 테이블을 기반으로 조인한다.
- 하나의 테이블이 메모리에 온전히 들어간다면 보통 중첩 루프 조인보다 더 효율적이다.
- 메모리에 올릴 수 없을 정도로 크다면 디스크를 사용하는 비용 발생한다.
    
    **빌드단계**
    
    - 입력 테이블 중 하나를 기반으로 메모리 내 해시 테이블을 빌드하는 단계
    - **바이트가 더 작은 테이블**을 기반으로 해서  테이블을 빌드
    
    **프로브 단계**
    
    - 각 레코드에서 ‘persons.contry_id에 일치하는 레코드를 찾아서 결괏값으로 반환한다.(해시함수> 해시값> 매칭레코드)
    - 테이블은 한 번씩만 읽게 되어 중첩해서 두 개의 테이블을 읽는 중첩루프 조인보다 보통 성능이 더 좋다.
    - 사용 가능한 메모리양은 시스템변수 join_buffer_size에 의해 제어되며, 런타임 시에 조정 가능하다.

**Comment by jindream6128** [comment]
조인은.. 프로그래머스 문제푸는게 짱짱맨 이었던거 같아용!

**Comment by 24tngus** [comment]
확인했습니다

**Comment by berryberries** [comment]
확인했습니다!
